### PR TITLE
[PLAT-167] SQL : imlement BETWEEN operator support

### PR DESCRIPTION
**Pull request by Fly-Style**
This PR implements BETWEEN operator support for Hazelcast SQL engine.

BETWEEN operator has two modes : 
- `ASYMMETRIC`,  default mode. It transforms statement `a BETWEEN b AND c` to the simple `a >= b AND a <= c` conjuction comparisons.
- `SYMMETRIC`. It transforms statement `a BETWEEN SYMMETRIC b AND c` to way more complex statement : 
`(a >= b AND a <= c) OR (a >= c AND a <= b)`.
Implementation contains these transformations in [HazelcastSqlToRel](https://github.com/hazelcast/hazelcast/pull/18483/files#diff-2f8029f211af628d820bc893a8fc8e64543403ca20d3c8eaad514672748e9143R227) class.

Operator grammar:
``` 
<between predicate> ::= <row value expression> [ NOT ] BETWEEN [ ASYMMETRIC | SYMMETRIC ]
                        <lower value expression> AND <upper value expression>
```
Checklist:
- [x] Labels and Milestone set

To know more about `SYMMETRIC` specificity : [here](https://blog.jooq.org/tag/between-symmetric/).

### CODE REVIEW

**Comment by Fly-Style** [comment]
run-lab-run

**Comment by Fly-Style** [comment]
run-lab-run

File: hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/predicate/HazelcastBetweenOperator.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+    private static final String[] BETWEEN_NAMES = {"BETWEEN ASYMMETRIC", "AND"};
+    private static final String[] NOT_BETWEEN_NAMES = {"NOT BETWEEN ASYMMETRIC", "AND"};
+    private static final String[] SYMMETRIC_BETWEEN_NAMES = {"BETWEEN SYMMETRIC", "AND"};
+    private static final String[] SYMMETRIC_NOT_BETWEEN_NAMES = {"NOT BETWEEN SYMMETRIC", "AND"};
+
+    private static final int PRECEDENCE = 32;
+    private static final int OPERANDS = 3;
+
+    static {
```
Comment (gierlachg) [inline hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/predicate/HazelcastBetweenOperator.java:54]: Why not inline?

File: hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/predicate/HazelcastBetweenOperator.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        BETWEEN_ASYMMETRIC = new HazelcastBetweenOperator(false, Flag.ASYMMETRIC);
+        NOT_BETWEEN_ASYMMETRIC = new HazelcastBetweenOperator(true, Flag.ASYMMETRIC);
+        BETWEEN_SYMMETRIC = new HazelcastBetweenOperator(false, Flag.SYMMETRIC);
+        NOT_BETWEEN_SYMMETRIC = new HazelcastBetweenOperator(true, Flag.SYMMETRIC);
+    }
+
+    private final boolean negated;
+    private final Flag flag;
+
+    protected HazelcastBetweenOperator(boolean negated, Flag symmetricalFlag) {
```
Comment (gierlachg) [inline hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/predicate/HazelcastBetweenOperator.java:64]: Can be `private`

File: hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/HazelcastSqlToRelConverter.java [hunk 3]
```diff
@@ -206,6 +220,50 @@ private RexNode convertCast(SqlCall call, Blackboard blackboard) {
         return getRexBuilder().makeCast(to, convertedOperand);
     }
 
+    /**
+     * Convert "val BETWEEN lower_bound AND upper_bound" expression to
+     * 1. If ASYMMETRIC : val >= lower_bound AND val =< upper_bound " expression. Default mode.
+     * 2. If SYMMETRIC : (val >= lower_bound AND val =< upper_bound) OR (val <= lower_bound AND val >= upper_bound)
+     * expression
+     */
+    public RexNode convertBetween(SqlCall call, Blackboard blackboard) {
```
Comment (gierlachg) [inline hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/HazelcastSqlToRelConverter.java:229]: Could not spot any differences between(!) this and what is in `StandardConvertletTable.convertBetween()`. Are there any?

Comment (Fly-Style) [reply hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/HazelcastSqlToRelConverter.java:229]: Yes, they are.` StandardConvertletTable#convertBetween` has a signature 
`public RexNode convertBetween(SqlRexContext cx, SqlBetweenOperator op, SqlCall call)`, which doesn't fit for us, because it uses `SqlBetweenOperator` as a second parameter. 
Our `HazelcastBetweenOperator` doesn't extends Calcite's `SqlBetweenOperator`.

File: hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/predicate/BetweenOperatorIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        checkValues(sqlQuery("BETWEEN ? AND ?"), INTEGER, new Integer[]{0, 1, 25}, 0, 25);
+        checkValues(sqlQuery("NOT BETWEEN ? AND ?"), INTEGER, new Integer[]{0, 1, 25, 30}, 25, 0);
+    }
+
+    @Test
+    public void betweenPredicateNullcheckTest() {
+        putAll(0, 1, 25, 30);
+        checkValues(sqlQuery("BETWEEN NULL AND 2"), INTEGER, new Integer[]{});
+        checkValues(sqlQuery("BETWEEN 2 AND NULL"), INTEGER, new Integer[]{});
+        checkValues(sqlQuery("BETWEEN NULL AND NULL"), INTEGER, new Integer[]{});
```
Comment (viliam-durina) [inline hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/predicate/BetweenOperatorIntegrationTest.java:69]: We should also test `NULL` on the left side: `null BETWEEN 1 AND 2`

File: hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/predicate/BetweenOperatorIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+    @Test
+    public void betweenPredicateDifferentTypesTest() {
+        LocalDate date = LocalDate.of(2000, 1, 1);
+        LocalTime time = LocalTime.of(8, 0, 0);
+        putAll(time, 1, "Argentina", "Bulgaria", "Ukraine", 2, date);
+
+        checkFailure0(sqlQuery("BETWEEN 1 AND 2"),
+            SqlErrorCode.DATA_EXCEPTION,
+            "Cannot parse VARCHAR value to BIGINT"
```
Comment (viliam-durina) [inline hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/predicate/BetweenOperatorIntegrationTest.java:168]: We should also test cases that pass, e.g. when the VARCHAR can actually be parsed into a BIGINT.

**Comment by Fly-Style** [comment]
run-lab-run

File: hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/predicate/HazelcastBetweenOperator.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        if (type.getComparability().ordinal() < RelDataTypeComparability.ALL.ordinal()) {
+            if (throwOnFailure) {
+                throw callBinding.newValidationSignatureError();
+            } else {
+                return false;
+            }
+        } else {
+            return true;
+        }
+    }
```
Comment (alex-dukhno) [inline hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/predicate/HazelcastBetweenOperator.java:116]:
```suggestion
    @Override
    protected boolean checkOperandTypes(HazelcastCallBinding callBinding, boolean throwOnFailure) {
        assert callBinding.getOperandCount() == OPERANDS;
        for (int i = 0; i < OPERANDS; ++i) {
            RelDataType type = callBinding.getOperandType(i);
            if (type.getComparability().ordinal() < RelDataTypeComparability.ALL.ordinal()) {
                if (throwOnFailure) {
                    throw callBinding.newValidationSignatureError();
                } else {
                    return false;
                }
            }
        }
        return true;
    }

    public Flag getFlag() {
        return flag;
    }

    public boolean isNegated() {
        return negated;
    }
```

Comment (Fly-Style) [reply hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/predicate/HazelcastBetweenOperator.java:116]: Good catch, thanks!

File: hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/HazelcastSqlToRelConverter.java [hunk 7]
```diff
@@ -250,6 +308,66 @@ private RexNode convertCall(SqlNode node, Blackboard blackboard) {
...
+    private static List<RexNode> convertExpressionList(RexBuilder rexBuilder,
+                                                       Blackboard bb,
+                                                       List<SqlNode> nodes,
+                                                       SqlOperandTypeChecker.Consistency consistency
+    ) {
+        final List<RexNode> exprs = new ArrayList<>();
+        for (SqlNode node : nodes) {
+            exprs.add(bb.convertExpression(node));
+        }
+        if (exprs.size() > 1) {
```
[the remark this answers is not in the document]
Comment (Fly-Style) [reply hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/HazelcastSqlToRelConverter.java:320]: Good catch, thanks!

File: hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/predicate/HazelcastBetweenOperator.java (hunk 1 above)
Comment (Fly-Style) [reply hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/predicate/HazelcastBetweenOperator.java:53]: It is more readable in this concrete case(IMO)

Comment (viliam-durina) [reply hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/predicate/HazelcastBetweenOperator.java:53]: @gierlachg It's because of checkstyle - `BETWEEN_NAMES` is private and you can't move it before `BETWEEN_ASYMMETRIC`, which is public 🤷

File: hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/predicate/BetweenOperatorIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+import static com.hazelcast.sql.SqlColumnType.TIMESTAMP_WITH_TIME_ZONE;
+import static com.hazelcast.sql.SqlColumnType.VARCHAR;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+
+
+/**
+ * <p> Test organization for BETWEEN operator.
+ * <p> Hazelcast SQL engine has 13 types in their type system (also OBJECT and NULL, but they don't count in case of BETWEEN operator)
```
Comment (alex-dukhno) [inline hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/predicate/BetweenOperatorIntegrationTest.java:58]: what if `OBJECT` is `Comparable`? see [ComparisonPredicate ](https://github.com/hazelcast/hazelcast/blob/fb720d9e7f4b7643c31fdb61e9b83b6ca8ecc552/hazelcast/src/main/java/com/hazelcast/sql/impl/expression/predicate/ComparisonPredicate.java#L82) for example

File: hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/predicate/BetweenOperatorIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+            List<SqlRow> rows = execute(member, sql, params);
+            for (SqlRow row : rows) {
+                // Here we have tested ComparisonPredicate, so we just check the type correctness.
+                assertEquals(expectedType, row.getMetadata().getColumn(0).getType());
+            }
+            return tuple2(rows, null);
+        } catch (HazelcastSqlException e) {
+            return tuple2(Collections.emptyList(), e);
+        } finally {
+            assertEquals("Impossible situation, not reachable", 0, 0);
```
Comment (viliam-durina) [inline hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/predicate/BetweenOperatorIntegrationTest.java:304]: What's the point of this?

File: hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/predicate/BetweenOperatorIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+                assertEquals(expectedObject, actualObject);
+
+                SqlColumnType expectedType = expectedResultsList.get(i).getMetadata().getColumn(0).getType();
+                SqlColumnType actualType = rows.get(i).getMetadata().getColumn(0).getType();
+                assertEquals(expectedType, actualType);
+            }
+        } catch (HazelcastSqlException e) {
+            assertNotNull(expectedOutcome.f1());
+            // Expected : ... Parameter at position 0 must be of BIGINT type, but VARCHAR was found
+            // Actual   : ... Parameter at position 0 must be of TINYINT type, but VARCHAR was found
```
Comment (viliam-durina) [inline hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/predicate/BetweenOperatorIntegrationTest.java:283]: This actually is an issue, it has to say "BIGINT" in both cases. I added a specific test just for this. Basically we need to replicate [this behavior](https://github.com/hazelcast/hazelcast/blob/master/hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/BinaryOperatorOperandTypeInference.java#L55-L66) of binary operator.

File: hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/predicate/BetweenOperatorIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+ *
+ * @see #betweenAsymmetricPredicateTypeCheckTest()
+ * @see #betweenSymmetricPredicateTypeCheckTest()
+ */
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
+public class BetweenOperatorIntegrationTest extends ExpressionTestSupport {
+
+    // TODO [viliam] remove OBJECT?
+    static final ExpressionType<?>[] TESTED_TYPES = ExpressionTypes.allExcept(ExpressionTypes.OBJECT);
```
Comment (viliam-durina) [inline hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/predicate/BetweenOperatorIntegrationTest.java:94]: We probably should not ignore OBJECT, the behavior should be the same in this case too.

**Review by viliam-durina — CHANGES_REQUESTED** [review]

File: hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/predicate/BetweenOperatorIntegrationTest.java (hunk 10 above)
Comment (Fly-Style) [reply hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/predicate/BetweenOperatorIntegrationTest.java]: Done.

**Review by viliam-durina — APPROVED** [review]

**Review by alex-dukhno — APPROVED** [review]
